src/ConnectionPool.jl

```diff
@@ -237,12 +250,19 @@ end
 Wait for prior pending writes to complete.
 """
 function IOExtras.startwrite(t::Transaction)
-    @require !iswritable(t)                     ;t.c.writecount != t.sequence &&
-                                                   @debug 1 "⏳  Wait write: $t"
-    while t.c.writecount != t.sequence
-        wait(t.c.writedone)
-    end                                           ;@debug 2 "👁  Start write:$t"
-    t.c.writebusy = true
+    @require !iswritable(t)
+
+    writebusy(t.c) && @debug 1 "⏳  Wait write: $t"
+    while true
+        lock = take!(t.c.writelock)
+        if t.c.writecount == t.sequence
```

Comment:
I do wonder whether `writecount` and `readcount` should be `Threads.Atomic{Int}` on the `Connection` type; they're only ever updated by a `Transaction` who holds the corresponding `writelock`/`readlock`, but they could be accessed by multiple transactions who are all waiting to pipeline a request through. In my current mental model, this should be fine since the field "updating" is locked, while reading is a free-for-all, but I don't know if there's some kind of potential issue/danger of, say, multiple threads all accessing a single struct's Int field value at the same time.

Reply:
For some really nice commentary on this, check out yuyichao's post here:
https://github.com/JuliaGPU/CUDAapi.jl/issues/106#issuecomment-585568882
Also look at the article linked at the top.

As a very rough summary: the problem with non-atomic access on the reader side is the reader may see something valid-looking for the variable, but even so, they can't use that to make assumptions about the rest of the data structures in memory which have been touched by the writer thread: the order in which those writes are done can look completely random to the reader even though they are *apparently sequential* in the writer thread.

This is a fact about compilers, but also a fundamental fact about computer hardware which can buffer writes in a cache before flushing them out for other cores to see.

If you have one writer and multiple readers and you are in doubt at all, just use `Atomic` or a lock. If you don't do this there are many very subtle ways things can go wrong. (Perhaps if you don't rely on the value that the readers get — ie, you never take a branch or do a load based on this value you might be ok. But then, why are you reading the value :-) )

Reply:
That makes a lot of sense. Thanks for the explanation!

Reply:
By the way, there are also more obvious problems with non-atomic loads for variables which are wider than the machine's native width or misaligned in a way that particular hardware doesn't like: in those cases it's possible to see a partial update (I believe this is what yuyichao is referring to as a "torn read" in the other thread).

Reply:
machine's native width isn't enough. Only atomic is atomic.
Modern CPUs are wild, and anything can be nonatomic if it feels like it.
There are no promises that you won't see torn reads of 32  bit values on a 32 bit system.

Reply:
> anything can be nonatomic if it feels like it

Well I don't think it's so bad as that; normal writes still need to go through the some of the same hardware so there will be some guarantees better than "all bets are off" (the following article suggests word size and alignment are the main things https://dave.cheney.net/2018/01/06/if-aligned-memory-writes-are-atomic-why-do-we-need-the-sync-atomic-package).

But regardless of that, "just use `Atomic`" is the right option for high level source code.